File: items/services/items.py
# module imports
from data.repositories.implementation.items_repository import itemsRepository
from api.routers.request_models.items import *
from api.routers.request_models.account import Account
from api.routers.error_handlers import *
from utils.utils import *
from config import Config
from .billing import BillingUnits

# python imports
from typing import *
import json

# external imports
from fastapi import Request


class ItemsService():
    """
        ItemsService
        Handles:
        - Get Items
        - Items creation
        - Update Items
        - Delete Items
        ...

        Attributes
        ----------
        account : Octy account
    """
    def __init__(self, account : Account, account_id : str = None): 
        # self.account = account
        # self.account_id = account_id if account_id != None else account.account_id
        # self.b = None if self.account is None else BillingUnits(account_id=self.account.account_id, account_type=self.account.account_configurations['a_t'], account_currency=self.account.account_configurations['a_c'], process_name='items_data')
        self.account = account
        if account_id is not None:
            self.account_id = account_id
        elif account is not None:
            self.account_id = account.account_id
    
        if self.account is None:
           self.b = None 
        else:
           self.b = BillingUnits(
            account_id=self.account.account_id, 
            account_type=self.account.account_configurations['a_t'], 
            account_currency=self.account.account_configurations['a_c'], 
            process_name='items_data'
        )

    def get_items(self,
                  item_ids : list = None, 
                  cursor : int = None) -> Union[dict, int]:
        """
        Parameters
        ----------
        item_ids : list
            list of item_ids
        cursor : int
            Pagination cursor

        Returns
        ----------
        items : dict
        total : int
        """
        if item_ids != None and cursor == 0:
            items = itemsRepository.get_item_by_ids(item_ids=item_ids,account_id=self.account.account_id)
            count = len(items)
            if count<1:
                raise OctyException(400, 'Invalid item identifier(s) provided', 
                [{'error_message' : 'No items were found with the provided identifier(s)', 
                'extended_help': Config['ITEMS_EXTENDED_HELP']}])
            
            return items, count
            

        elif item_ids == None and cursor != None:
            
            items,total = itemsRepository.get_items(account_id=self.account.account_id, 
                                                cursor=cursor)
            if len(items)<1:
                raise OctyException(400, 'No items found', 
                [{'error_message' : 'No items found with the provided item identifier or pagination cursor exhausted', 
                'extended_help': Config['ITEMS_EXTENDED_HELP']}])
            return items, total

    async def create_items(self, items : CreateItems) -> Union[list, list]:
        """
        Parameters
        ----------
        items : CreateItems
            CreateItems request model instance

        Returns
        ----------
        Created and failed to create items : Union[list, list]
        """

        # assess allowed limits
        res, counts = assess_resource_limit(self.account.account_configurations['li'],
                              itemsRepository.get_item_count(self.account.account_id),
                              len(items.items))
        if not res:
            raise OctyException(400,'Resource limit exceeded', 
            [{'error_message' : f'This request could not be completed as the number of items sent with this request exceeds the allowed limit of : {counts["limit"]}. This account can create another {counts["remainder"]} items.', 'extended_help': Config['RATE_LIMIT_EXTENDED_HELP']}])

        items_batch = []
        for item in items.items:
            items_batch.append(
                {
                    'item_id' : item.item_id,
                    'account_id' : self.account.account_id,
                    'item_category' : item.item_category,
                    'item_name' : item.item_name,
                    'item_description' : item.item_description,
                    'item_price' : item.item_price,
                    'event_type' : 'charged'
                }
            )

        created, failed = itemsRepository.create_items(items_batch)

        if len(created) < 1:
            raise OctyException(400, 'No items created!', failed)

        await self.b.track_data_units(created)
        await self.b.complete_data_units('MB')

        return created, failed
    
    async def update_items(self, items : UpdateItems) -> Union[list, list]:
        """
        Parameters
        ----------
        items : UpdateItems
            UpdateItems request model instance

        Returns
        ----------
        Updated and failed to update items : Union[list, list]
        """
        items_batch = []
        for item in items.items:
            items_batch.append(
                {
                    'item_id' : item.item_id,
                    'account_id' : self.account.account_id,
                    'item_category' : item.item_category,
                    'item_name' : item.item_name,
                    'item_description' : item.item_description,
                    'item_price' : item.item_price,
                    'status' : item.status,
                    'event_type' : 'charged'
                }
            )

        updated, failed = itemsRepository.update_items(items_batch, self.account.account_id)

        if len(updated) < 1:
            raise OctyException(400, 'No items updated!', failed)

        await self.b.track_data_units(updated)
        await self.b.complete_data_units('MB')

        return updated, failed

    async def delete_items(self, items : DeleteItems) -> Union[list, list]:
        """
        Parameters
        ----------
        items : DeleteItems
            DeleteItems request model instance
    
        Returns
        ----------
        Deleted and failed to delete item ids : Union[list, list]
        """
        items_batch = []
        for item in items.items:
            items_batch.append({
                "item_id" : item,
                "account_id" : self.account.account_id
            })

        deleted , failed = await itemsRepository.delete_items(items_batch, self.account)

        if len(deleted) < 1:
            raise OctyException(400, 'No items deleted!', failed)
        return deleted, failed


    #INTERNAL

    def get_items_internal(self, account_id : str, cursor : int, ids : bool, status : str) -> Union[dict, int]:
        """
        Parameters
        ----------
        account_id : str
            Octy account id
        cursor : int
            Pagination cursor
        ids : bool
            Only get item id(s)
        status : str

        Returns
        ----------
        items : dict
        total : int
        """
        items,total = itemsRepository.get_items(account_id=account_id, cursor=cursor, ids=ids, status=status)
        if len(items)<1:
            raise OctyException(400, 'No items found', 
            [{'error_message' : 'No items found or pagination cursor exhausted', 
            'extended_help': Config['ITEMS_EXTENDED_HELP']}])
        return items, total
    
    #Delete all items for an account
    async def delete_account_items_internal(self, account_id : str) -> bool:
        """
        Parameters
        ----------
        account_id : str
            Octy account id

        Returns
        ----------
        bool
        """
        res = await itemsRepository.delete_account_items_internal(account_id=account_id)
        return res

File: rust/services/items/src/services/items.rs
//! Port of `services/items.py` (`ItemsService`).

use octy_shared::errors::{ErrorReason, OctyError};
use octy_spin::auth::AuthAccount;
use octy_spin::ctx::Ctx;
use serde_json::{json, Value};

use crate::errors::ApiError;
use crate::models::{CreateItems, DeleteItems, UpdateItems};
use crate::repos::items as repo;
use crate::services::billing::BillingUnits;

pub struct ItemsService<'a> {
    ctx: &'a Ctx,
    account: Option<&'a AuthAccount>,
    billing: Option<BillingUnits>,
}

/// `utils.assess_resource_limit` — index 1 of the `li` limits string is the
/// items limit.
struct LimitCounts {
    limit: i64,
    remainder: i64,
}

fn assess_resource_limit(
    limits: &str,
    current_count: i64,
    requested: i64,
) -> Result<(bool, LimitCounts), OctyError> {
    let raw = limits
        .split('*')
        .nth(1)
        .ok_or_else(|| OctyError::internal("IndexError: list index out of range"))?;
    let resource_limit: i64 = raw.trim().parse().map_err(|_| {
        OctyError::internal(format!("invalid resource limit value: {raw}"))
    })?;
    let remainder = resource_limit - current_count;

    if requested + current_count > resource_limit {
        Ok((
            false,
            LimitCounts {
                limit: resource_limit,
                remainder,
            },
        ))
    } else {
        Ok((
            true,
            LimitCounts {
                limit: resource_limit,
                remainder: remainder - requested,
            },
        ))
    }
}

impl<'a> ItemsService<'a> {
    /// `ItemsService(account)` — the Python constructor eagerly built
    /// `BillingUnits` from `a_cf['a_t']` / `a_cf['a_c']` (KeyError → 500).
    pub fn new(ctx: &'a Ctx, account: &'a AuthAccount) -> Result<Self, OctyError> {
        let billing = BillingUnits::for_account(account, "items_data")?;
        Ok(Self {
            ctx,
            account: Some(account),
            billing: Some(billing),
        })
    }

    /// `ItemsService(None)` / `ItemsService(account_id=…)` — internal routes.
    pub fn internal(ctx: &'a Ctx) -> Self {
        Self {
            ctx,
            account: None,
            billing: None,
        }
    }

    fn items_help(&self) -> Result<&str, OctyError> {
        self.ctx.config.get_str("ITEMS_EXTENDED_HELP")
    }

    fn account(&self) -> &AuthAccount {
        self.account.expect("authenticated route")
    }

    pub async fn get_items(
        &mut self,
        item_ids: Option<&[String]>,
        cursor: i64,
    ) -> Result<(Vec<Value>, i64), ApiError> {
        let account_id = self.account().account_id.clone();

        if let Some(ids) = item_ids {
            if cursor == 0 {
                let items = repo::get_item_by_ids(self.ctx, ids, &account_id).await?;
                let count = items.len() as i64;
                if count < 1 {
                    return Err(OctyError::new(
                        400,
                        "Invalid item identifier(s) provided",
                        vec![ErrorReason::new(
                            "No items were found with the provided identifier(s)",
                            self.items_help()?,
                        )],
                    )
                    .into());
                }
                return Ok((items, count));
            }
        } else {
            let (items, total) =
                repo::get_items(self.ctx, &account_id, cursor, false, "all").await?;
            if items.len() < 1 {
                return Err(OctyError::new(
                    400,
                    "No items found",
                    vec![ErrorReason::new(
                        "No items found with the provided item identifier or pagination cursor exhausted",
                        self.items_help()?,
                    )],
                )
                .into());
            }
            return Ok((items, total));
        }

        // Python fell through returning None (item_ids with nonzero cursor
        // never happens — the router always passes cursor=0 with ids) and the
        // DTO then crashed with a TypeError → 500.
        Err(OctyError::internal("TypeError: cannot unpack non-iterable NoneType object").into())
    }

    pub async fn create_items(
        &mut self,
        items: CreateItems,
    ) -> Result<(Vec<Value>, Vec<Value>), ApiError> {
        let account = self.account();
        let account_id = account.account_id.clone();

        // assess allowed limits
        let li = account
            .account_configurations
            .get("li")
            .and_then(Value::as_str)
            .ok_or_else(|| OctyError::internal("KeyError: 'li'"))?
            .to_string();
        let current_count = repo::get_item_count(self.ctx, &account_id).await?;
        let (res, counts) = assess_resource_limit(&li, current_count, items.items.len() as i64)?;
        if !res {
            let rate_help = self.ctx.config.get_str("RATE_LIMIT_EXTENDED_HELP")?;
            return Err(OctyError::new(
                400,
                "Resource limit exceeded",
                vec![ErrorReason::new(
                    format!(
                        "This request could not be completed as the number of items sent with this request exceeds the allowed limit of : {}. This account can create another {} items.",
                        counts.limit, counts.remainder
                    ),
                    rate_help,
                )],
            )
            .into());
        }

        let items_batch: Vec<Value> = items
            .items
            .iter()
            .map(|item| {
                json!({
                    "item_id": item.item_id,
                    "account_id": account_id,
                    "item_category": item.item_category,
                    "item_name": item.item_name,
                    "item_description": item.item_description,
                    "item_price": item.item_price,
                    "event_type": "charged",
                })
            })
            .collect();

        let (created, failed) = repo::create_items(self.ctx, &items_batch).await?;

        if created.len() < 1 {
            return Err(ApiError::raw(400, "No items created!", failed));
        }

        let billing = self.billing.as_mut().expect("billing on authenticated route");
        billing.track_data_units(&json!(created));
        billing.complete_data_units(self.ctx, "MB").await?;

        Ok((created, failed))
    }

    pub async fn update_items(
        &mut self,
        items: UpdateItems,
    ) -> Result<(Vec<Value>, Vec<Value>), ApiError> {
        let account_id = self.account().account_id.clone();

        let items_batch: Vec<Value> = items
            .items
            .iter()
            .map(|item| {
                json!({
                    "item_id": item.item_id,
                    "account_id": account_id,
                    "item_category": item.item_category,
                    "item_name": item.item_name,
                    "item_description": item.item_description,
                    "item_price": item.item_price,
                    "status": item.status,
                    "event_type": "charged",
                })
            })
            .collect();

        let items_help = self.items_help()?.to_string();
        let (updated, failed) =
            repo::update_items(self.ctx, &items_batch, &account_id, &items_help).await?;

        if updated.len() < 1 {
            return Err(ApiError::raw(400, "No items updated!", failed));
        }

        let billing = self.billing.as_mut().expect("billing on authenticated route");
        billing.track_data_units(&json!(updated));
        billing.complete_data_units(self.ctx, "MB").await?;

        Ok((updated, failed))
    }

    pub async fn delete_items(
        &mut self,
        items: DeleteItems,
    ) -> Result<(Vec<Value>, Vec<Value>), ApiError> {
        let account = self.account();
        let account_id = account.account_id.clone();

        let items_batch: Vec<Value> = items
            .items
            .iter()
            .map(|item| {
                json!({
                    "item_id": item,
                    "account_id": account_id,
                })
            })
            .collect();

        let (deleted, failed) = repo::delete_items(self.ctx, &items_batch, account).await?;

        if deleted.len() < 1 {
            return Err(ApiError::raw(400, "No items deleted!", failed));
        }
        Ok((deleted, failed))
    }

    // ---- INTERNAL ----

    pub async fn get_items_internal(
        &mut self,
        account_id: &str,
        cursor: i64,
        ids: bool,
        status: &str,
    ) -> Result<(Vec<Value>, i64), ApiError> {
        let (items, total) =
            repo::get_items(self.ctx, &json!(account_id), cursor, ids, status).await?;
        if items.len() < 1 {
            return Err(OctyError::new(
                400,
                "No items found",
                vec![ErrorReason::new(
                    "No items found or pagination cursor exhausted",
                    self.items_help()?,
                )],
            )
            .into());
        }
        Ok((items, total))
    }

    /// Delete all items for an account (account-deletion fan-out).
    pub async fn delete_account_items_internal(&mut self, account_id: &str) -> Result<bool, ApiError> {
        Ok(repo::delete_account_items_internal(account_id).await?)
    }
}
